File: cwclock-mobile/lib/screens/time/time_tracker_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../common/format.dart';
import '../../common/project_label.dart';
import '../../models/time_entry.dart';
import '../../providers/clients_provider.dart';
import '../../providers/locale_provider.dart';
import '../../providers/projects_provider.dart';
import '../../providers/session_provider.dart';
import '../../providers/time_entries_provider.dart';
import '../../theme.dart';
import '../../widgets/app_button.dart';
import '../../widgets/form_field.dart';
import '../../widgets/select_field.dart';

/// The default tab: a running-timer header (start/stop) above a
/// day-grouped list of entries. All-day records are created from
/// AllDayRecordScreen (the "+" button); tapping a row opens EditRecordScreen.
/// Ported from src/screens/time/TimeTrackerScreen.js.
class TimeTrackerScreen extends ConsumerStatefulWidget {
  const TimeTrackerScreen({super.key});

  @override
  ConsumerState<TimeTrackerScreen> createState() => _TimeTrackerScreenState();
}

class _TimeTrackerScreenState extends ConsumerState<TimeTrackerScreen> {
  final _scrollController = ScrollController();
  String _projectId = '';
  String _text = '';
  int _elapsed = 0;
  bool _starting = false;
  Timer? _ticker;

  @override
  void initState() {
    super.initState();
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId != null) {
      Future.microtask(() {
        ref.read(projectsProvider.notifier).listProjects(orgId);
        ref.read(clientsProvider.notifier).listClients(orgId);
        ref.read(timeEntriesProvider.notifier).listTimeEntries(orgId, 1);
        ref.read(runningTimerProvider.notifier).load();
      });
    }
    _scrollController.addListener(_onScroll);
  }

  void _onScroll() {
    if (!_scrollController.hasClients) return;
    final threshold = _scrollController.position.maxScrollExtent * 0.6;
    if (_scrollController.position.pixels >= threshold) {
      final entriesState = ref.read(timeEntriesProvider);
      final orgId = ref.read(sessionProvider).orgId;
      if (orgId != null && entriesState.hasMore && !entriesState.isLoadingMore) {
        ref.read(timeEntriesProvider.notifier).listTimeEntries(orgId, entriesState.page + 1);
      }
    }
  }

  void _startTicking(DateTime startedAt) {
    _ticker?.cancel();
    void tick() {
      if (!mounted) return;
      final diff = DateTime.now().difference(startedAt).inSeconds;
      setState(() {
        _elapsed = diff < 0 ? 0 : diff;
      });
    }

    tick();
    _ticker = Timer.periodic(const Duration(seconds: 1), (_) => tick());
  }

  @override
  void dispose() {
    _ticker?.cancel();
    _scrollController.dispose();
    super.dispose();
  }

  Future<void> _handleStart() async {
    if (_projectId.isEmpty) return;
    final projects = ref.read(projectsProvider).items;
    final project = projects.where((p) => p.id == _projectId).firstOrNull;
    if (project == null) return;

    setState(() => _starting = true);
    await ref.read(runningTimerProvider.notifier).start(
      projectId: project.id,
      clientId: project.clientId,
      text: _text.trim().isEmpty ? project.name : _text.trim(),
    );
    setState(() => _starting = false);
  }

  Future<void> _handleStop() async {
    final timer = ref.read(runningTimerProvider);
    if (timer == null) return;
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;
    final started = DateTime.parse(timer.startedAt);
    final now = DateTime.now();
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);

    try {
      await ref.read(timeEntriesProvider.notifier).createTimeEntry(orgId, {
        'text': timer.text,
        'day': toDayString(started),
        'start': toHMS(started),
        'end': toHMS(now),
        'allDay': false,
        'half': false,
        'clientId': timer.clientId,
        'projectId': timer.projectId,
      });
      await ref.read(runningTimerProvider.notifier).clear();
      _ticker?.cancel();
      setState(() {
        _text = '';
        _projectId = '';
      });
    } catch (e) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text(t('common.retry'))));
      }
    }
  }

  void _handleDelete(TimeEntry item) {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;

    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text(t('timeTracker.deleteRecordTitle')),
        content: Text(t('timeTracker.deleteRecordBody', {'text': item.text})),
        actions: [
          TextButton(onPressed: () => Navigator.pop(context), child: Text(t('common.cancel'))),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ref.read(timeEntriesProvider.notifier).deleteTimeEntry(orgId, item.id);
            },
            child: Text(t('common.delete'), style: TextStyle(color: AppColors.of(context).danger)),
          ),
        ],
      ),
    );
  }

  String _timeLabel(TimeEntry item, String Function(String, [Map<String, String>?]) t) {
    if (item.allDay) return t('timeTracker.allDay');
    if (item.half) return t('timeTracker.half');
    return '${padTimeString(item.start).isEmpty ? '?' : padTimeString(item.start)} - '
        '${padTimeString(item.end).isEmpty ? '?' : padTimeString(item.end)}';
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final session = ref.watch(sessionProvider);
    final projects = ref.watch(projectsProvider).items;
    final clients = ref.watch(clientsProvider).items;
    final entriesState = ref.watch(timeEntriesProvider);
    final timer = ref.watch(runningTimerProvider);

    ref.listen(runningTimerProvider, (previous, next) {
      if (next != null) _startTicking(DateTime.parse(next.startedAt));
    });

    final projectItems = [
      for (final p in projects) SelectItem(p.id, projectLabel(p, clients)),
    ];

    final groups = groupByDay(entriesState.items);

    return Scaffold(
      appBar: AppBar(title: Text(t('timeTracker.title'))),
      body: Column(
        children: [
          Container(
            padding: EdgeInsets.all(AppSpacing.of(2)),
            decoration: BoxDecoration(
              border: Border(bottom: BorderSide(color: AppColors.of(context).border)),
            ),
            child: timer != null
                ? Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        timer.text,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: TextStyle(fontSize: 16, fontWeight: FontWeight.w600, color: AppColors.of(context).text),
                      ),
                      Padding(
                        padding: EdgeInsets.symmetric(vertical: AppSpacing.of(1)),
                        child: Text(
                          formatHMS(_elapsed),
                          style: TextStyle(fontSize: 28, fontWeight: FontWeight.w700, color: AppColors.of(context).primary),
                        ),
                      ),
                      AppButton(title: t('timeTracker.stop'), variant: AppButtonVariant.danger, onPressed: _handleStop),
                    ],
                  )
                : Column(
                    children: [
                      AppSelectField(
                        label: t('timeTracker.project'),
                        value: _projectId,
                        onChanged: (v) => setState(() => _projectId = v),
                        items: projectItems,
                        placeholder: t('timeTracker.project'),
                      ),
                      AppFormField(
                        value: _text,
                        onChanged: (v) => setState(() => _text = v),
                        placeholder: t('timeTracker.whatAreYouWorkingOn'),
                      ),
                      AppButton(
                        title: t('timeTracker.start'),
                        onPressed: _projectId.isEmpty ? null : _handleStart,
                        loading: _starting,
                      ),
                    ],
                  ),
          ),
          Expanded(
            child: Stack(
              children: [
                RefreshIndicator(
                  onRefresh: () async {
                    if (session.orgId != null) {
                      await ref.read(timeEntriesProvider.notifier).listTimeEntries(session.orgId!, 1);
                    }
                  },
                  child: groups.isEmpty && !entriesState.isLoading
                      ? ListView(
                          children: [
                            Padding(
                              padding: EdgeInsets.all(AppSpacing.of(3)),
                              child: Text(
                                t('timeTracker.noRecords'),
                                textAlign: TextAlign.center,
                                style: TextStyle(color: AppColors.of(context).textMuted),
                              ),
                            ),
                          ],
                        )
                      : ListView.builder(
                          controller: _scrollController,
                          padding: EdgeInsets.only(bottom: AppSpacing.of(10)),
                          itemCount: groups.fold<int>(0, (sum, g) => sum + 1 + g.items.length),
                          itemBuilder: (context, index) {
                            var remaining = index;
                            for (final group in groups) {
                              if (remaining == 0) {
                                return Container(
                                  color: AppColors.of(context).backgroundMuted,
                                  padding: EdgeInsets.symmetric(
                                    horizontal: AppSpacing.of(2),
                                    vertical: AppSpacing.of(1),
                                  ),
                                  child: Row(
                                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                    children: [
                                      Text(
                                        group.day,
                                        style: TextStyle(
                                            fontSize: 13, fontWeight: FontWeight.w600, color: AppColors.of(context).textMuted),
                                      ),
                                      Text(
                                        formatHMS(group.totalSecs),
                                        style: TextStyle(
                                            fontSize: 13, fontWeight: FontWeight.w600, color: AppColors.of(context).textMuted),
                                      ),
                                    ],
                                  ),
                                );
                              }
                              remaining -= 1;
                              if (remaining < group.items.length) {
                                final item = group.items[remaining];
                                final project = projects.where((p) => p.id == item.projectId).firstOrNull;
                                return ListTile(
                                  contentPadding: EdgeInsets.symmetric(horizontal: AppSpacing.of(2)),
                                  title: Text(item.text, maxLines: 1, overflow: TextOverflow.ellipsis),
                                  subtitle: Text(
                                    '${project != null ? projectLabel(project, clients) : ''} · ${_timeLabel(item, t)}',
                                    maxLines: 1,
                                    overflow: TextOverflow.ellipsis,
                                  ),
                                  trailing: TextButton(
                                    onPressed: () => _handleDelete(item),
                                    child: Text(t('common.delete'), style: TextStyle(color: AppColors.of(context).danger)),
                                  ),
                                  onTap: () => context.push('/edit-record', extra: item),
                                );
                              }
                              remaining -= group.items.length;
                            }
                            return const SizedBox.shrink();
                          },
                        ),
                ),
              ],
            ),
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => context.push('/all-day-record'),
        backgroundColor: AppColors.of(context).primary,
        child: const Icon(Icons.add, color: kWhite),
      ),
    );
  }
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull {
    final it = iterator;
    return it.moveNext() ? it.current : null;
  }
}

File: cwclock-mobile/src/screens/time/TimeTrackerScreen.js
import React, { useCallback, useEffect, useMemo, useState } from "react";
import { Alert, FlatList, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import Button from "../../components/Button";
import SelectField from "../../components/SelectField";
import FormField from "../../components/FormField";
import { useI18n } from "../../i18n/I18nContext";
import { apiErrorMessage } from "../../i18n/translate";
import theme from "../../theme";
import projectLabel from "../../common/projectLabel";
import { groupByDay, formatHMS, padTimeString, toDayString, toHMS } from "../../common/format";
import { listProjectsApi } from "../../redux/projects/projects.actions";
import { listClientsApi } from "../../redux/clients/clients.actions";
import {
  listTimeEntriesApi,
  createTimeEntryApi,
  deleteTimeEntryApi,
} from "../../redux/timeEntries/timeEntries.actions";
import { loadTimer, saveTimer, clearTimer } from "../../storage/timer";

// TimeTrackerScreen is the default tab: a running-timer header (start/stop,
// mirrors cwclock-ui's TaskInput.jsx) above a day-grouped list of entries
// (mirrors TasksApp.jsx's dayGroups). All-day records are created from
// AllDayRecordScreen (the "+" button); tapping a row opens EditRecordScreen.
const TimeTrackerScreen = ({ navigation }) => {
  const { t, locale } = useI18n();
  const dispatch = useDispatch();
  const { orgId } = useSelector((state) => state.session);
  const { items: projects } = useSelector((state) => state.projects);
  const { items: clients } = useSelector((state) => state.clients);
  const { items: entries, isLoading, isLoadingMore, hasMore, page } = useSelector((state) => state.timeEntries);

  const [projectId, setProjectId] = useState("");
  const [text, setText] = useState("");
  const [timer, setTimer] = useState(null);
  const [elapsed, setElapsed] = useState(0);
  const [starting, setStarting] = useState(false);

  useEffect(() => {
    if (!orgId) return;
    dispatch(listProjectsApi(orgId));
    dispatch(listClientsApi(orgId));
    dispatch(listTimeEntriesApi(orgId, 1));
    loadTimer().then(setTimer);
  }, [orgId, dispatch]);

  useEffect(() => {
    if (!timer) return undefined;
    const tick = () => setElapsed(Math.max(0, Math.floor((Date.now() - new Date(timer.startedAt).getTime()) / 1000)));
    tick();
    const id = setInterval(tick, 1000);
    return () => clearInterval(id);
  }, [timer]);

  const projectItems = useMemo(
    () => projects.map((p) => ({ value: p.id, label: projectLabel(p, clients) })),
    [projects, clients]
  );

  const handleStart = async () => {
    if (!projectId) return;
    const project = projects.find((p) => p.id === projectId);
    if (!project) return;
    const next = {
      startedAt: new Date().toISOString(),
      projectId: project.id,
      clientId: project.clientId,
      text: text.trim() || project.name,
    };
    setStarting(true);
    await saveTimer(next);
    setTimer(next);
    setStarting(false);
  };

  const handleStop = async () => {
    if (!timer) return;
    const started = new Date(timer.startedAt);
    const now = new Date();
    try {
      await dispatch(
        createTimeEntryApi(orgId, {
          text: timer.text,
          day: toDayString(started),
          start: toHMS(started),
          end: toHMS(now),
          allDay: false,
          half: false,
          clientId: timer.clientId,
          projectId: timer.projectId,
        })
      );
      await clearTimer();
      setTimer(null);
      setText("");
      setProjectId("");
    } catch (e) {
      Alert.alert(t("common.retry"), apiErrorMessage(e, locale));
    }
  };

  const handleDelete = (item) => {
    Alert.alert(t("timeTracker.deleteRecordTitle"), t("timeTracker.deleteRecordBody", { text: item.text }), [
      { text: t("common.cancel"), style: "cancel" },
      { text: t("common.delete"), style: "destructive", onPress: () => dispatch(deleteTimeEntryApi(orgId, item.id)) },
    ]);
  };

  const rows = useMemo(() => {
    const out = [];
    for (const group of groupByDay(entries)) {
      out.push({ rowType: "header", key: `h-${group.day}`, day: group.day, totalSecs: group.totalSecs });
      for (const item of group.items) {
        out.push({ rowType: "entry", key: item.id, item });
      }
    }
    return out;
  }, [entries]);

  const timeLabel = useCallback(
    (item) => {
      if (item.allDay) return t("timeTracker.allDay");
      if (item.half) return t("timeTracker.half");
      return `${padTimeString(item.start) || "?"} - ${padTimeString(item.end) || "?"}`;
    },
    [t]
  );

  const renderItem = ({ item: row }) => {
    if (row.rowType === "header") {
      return (
        <View style={styles.dayHeader}>
          <Text style={styles.dayHeaderText}>{row.day}</Text>
          <Text style={styles.dayHeaderText}>{formatHMS(row.totalSecs)}</Text>
        </View>
      );
    }
    const item = row.item;
    const project = projects.find((p) => p.id === item.projectId);
    return (
      <TouchableOpacity style={styles.entryRow} onPress={() => navigation.navigate("EditRecord", { entry: item })}>
        <View style={styles.entryMain}>
          <Text style={styles.entryText} numberOfLines={1}>
            {item.text}
          </Text>
          <Text style={styles.entrySub} numberOfLines={1}>
            {project ? projectLabel(project, clients) : ""} · {timeLabel(item)}
          </Text>
        </View>
        <TouchableOpacity onPress={() => handleDelete(item)} style={styles.deleteBtn}>
          <Text style={styles.deleteBtnText}>{t("common.delete")}</Text>
        </TouchableOpacity>
      </TouchableOpacity>
    );
  };

  return (
    <Screen scroll={false}>
      <View style={styles.header}>
        {timer ? (
          <>
            <Text style={styles.runningText} numberOfLines={1}>
              {timer.text}
            </Text>
            <Text style={styles.elapsed}>{formatHMS(elapsed)}</Text>
            <Button title={t("timeTracker.stop")} variant="danger" onPress={handleStop} />
          </>
        ) : (
          <>
            <SelectField
              label={t("timeTracker.project")}
              value={projectId}
              onValueChange={setProjectId}
              items={projectItems}
              placeholder={t("timeTracker.project")}
            />
            <FormField
              placeholder={t("timeTracker.whatAreYouWorkingOn")}
              value={text}
              onChangeText={setText}
            />
            <Button
              title={t("timeTracker.start")}
              onPress={handleStart}
              disabled={!projectId}
              loading={starting}
            />
          </>
        )}
      </View>

      <FlatList
        data={rows}
        keyExtractor={(row) => row.key}
        renderItem={renderItem}
        refreshing={isLoading}
        onRefresh={() => dispatch(listTimeEntriesApi(orgId, 1))}
        onEndReachedThreshold={0.4}
        onEndReached={() => {
          if (hasMore && !isLoadingMore) dispatch(listTimeEntriesApi(orgId, page + 1));
        }}
        ListEmptyComponent={!isLoading ? <Text style={styles.empty}>{t("timeTracker.noRecords")}</Text> : null}
        contentContainerStyle={styles.listContent}
      />

      <TouchableOpacity style={styles.fab} onPress={() => navigation.navigate("AllDayRecord")}>
        <Text style={styles.fabText}>+</Text>
      </TouchableOpacity>
    </Screen>
  );
};

const styles = StyleSheet.create({
  header: {
    padding: theme.spacing(2),
    borderBottomWidth: 1,
    borderBottomColor: theme.color.border,
  },
  runningText: {
    fontSize: 16,
    fontWeight: "600",
    color: theme.color.text,
  },
  elapsed: {
    fontSize: 28,
    fontWeight: "700",
    color: theme.color.primary,
    marginVertical: theme.spacing(1),
  },
  listContent: {
    paddingBottom: theme.spacing(10),
  },
  dayHeader: {
    flexDirection: "row",
    justifyContent: "space-between",
    paddingHorizontal: theme.spacing(2),
    paddingVertical: theme.spacing(1),
    backgroundColor: theme.color.backgroundMuted,
  },
  dayHeaderText: {
    fontSize: 13,
    fontWeight: "600",
    color: theme.color.textMuted,
  },
  entryRow: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: theme.spacing(2),
    paddingVertical: theme.spacing(1.5),
    borderBottomWidth: 1,
    borderBottomColor: theme.color.border,
  },
  entryMain: {
    flex: 1,
  },
  entryText: {
    fontSize: 16,
    color: theme.color.text,
  },
  entrySub: {
    fontSize: 13,
    color: theme.color.textMuted,
    marginTop: 2,
  },
  deleteBtn: {
    paddingHorizontal: theme.spacing(1),
  },
  deleteBtnText: {
    color: theme.color.danger,
    fontSize: 13,
  },
  empty: {
    padding: theme.spacing(3),
    textAlign: "center",
    color: theme.color.textMuted,
  },
  fab: {
    position: "absolute",
    right: theme.spacing(2),
    bottom: theme.spacing(2),
    width: 56,
    height: 56,
    borderRadius: 28,
    backgroundColor: theme.color.primary,
    alignItems: "center",
    justifyContent: "center",
    elevation: 4,
  },
  fabText: {
    color: theme.color.white,
    fontSize: 28,
    lineHeight: 30,
  },
});

export default TimeTrackerScreen;
